### DISCUSSION TITLE
Is there a way to make .padded().repeated() stop when encountering pattern?

### DISCUSSION THREAD

**Discussion by SergeyKasmy** [zesterer/chumsky#636] [Q&A]
Hello.

I'm looking for a way to make a `.padded().repeated()` parser stop at an empty new line, i.e. `newline().then(newline())` but sadly I haven't been successful.

For example, I have a file formatted like this:

```
HEADER1:
    stuff
    more_stuff

HEADER2: foo bar
```

At first, to parse HEADER1, I came up with something like this

```rust
let empty_line = newline().then(newline());

let header1= just(b"HEADER1:")
  .ignore_then(
    text::ident::<u8, Simple<u8>>()
      .padded()
      .repeated()
  );
```

Sadly this obviously eats up every other input after HEADER1.

Adding `.then_ignore(empty_line)` at the end doesn't work because `.repeated()` is eagerly evaluated.

I've tried replacing `.padded()` with `.padded_by(empty_line.not().or(whitespace()))` but it screams at me that `empty_line` and `whitespace()` parsers return different values.

Just making header1 the last parser in the chain isn't an option because I want to report any unrecognized lines in the file and not interpret them as part of header1, e.g. I want this to error:

```
HEADER2: foo bar

HEADER1:
    stuff

UNKNOWN_HEADER: bad
```

Is there anything I can do to make the `.padded().repeated()` parser stop at `\n\n`?

Thanks!

**Comment by zesterer** [comment]
You can do something like `a.and_is(b.not()).repeated()` to make the repeating pattern stop when it finds a `b`. This works even if `a` would normally successfully parse `b`.
